clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java

```diff
@@ -149,42 +133,44 @@ private void writeUnlockAll() {
     }
   }
 
-  private void handleOversizeMappingException(long key, LongConsumer operation) {
-    handleOversizeMappingException(key, k -> {
+  private void tryUntilSucceds(long key, LongConsumer operation) {
+    tryUntilSucceds(key, k -> {
       operation.accept(k);
       return null;
     });
   }
 
   /**
-   * Force eviction from other segments until {@code operation} succeeds or no further eviction is possible.
+   * Try to apply the {@code operation}. If failed with OversizeMappingException, then
+   * force eviction from other segments until {@code operation} succeeds or no further eviction is possible.
    *
    * @param key the target key
    * @param operation the previously failed operation
    * @param <R> operation result type
    * @return the operation result
    * @throws OversizeMappingException if the operation cannot be made to succeed
    */
-  private <R> R handleOversizeMappingException(long key, LongFunction<R> operation) throws OversizeMappingException {
-    if (tryShrinkOthers(key)) {
-      try {
-        return operation.apply(key);
-      } catch (OversizeMappingException ex) {
-        //ignore
-      }
+  private <R> R tryUntilSucceds(long key, LongFunction<R> operation) throws OversizeMappingException {
+    try {
+      return operation.apply(key);
+    } catch (OversizeMappingException e) {
+      return handleOversizeMappingException(key, operation, e);
     }
+  }
+
+  private <R> R handleOversizeMappingException(long key, LongFunction<R> operation, OversizeMappingException exception)
+    throws OversizeMappingException {
 
     writeLockAll();
```

Comment:
@chrisdennis .. One behaviour difference I see from previous version is that now the writelockall is attempted before a quick and dirty tryShrinkOthers..

Reply:
OK now I see the initial intent. The problem I am trying to fix here is that the failing quick and dirty tryShrinkOthers is followed by writeLockAll() and another call to operation.apply(key). This doesn't make sense to me because the shrinking has failed so there is no way the insertion will succeed. Am I wrong?